experiments/2017-09-21-rust-on-android.md

```diff
@@ -0,0 +1,224 @@
+
+Open `cargo/src/lib.rs`. At the bottom of the file add the following code:
+```
+/// Expose the JNI interface for android below
+#[cfg(target_os="android")]
+#[allow(non_snake_case)]
+pub mod android {
+    extern crate jni;
+
+    use super::*;
+    use self::jni::JNIEnv;
+    use self::jni::objects::{JClass, JString};
+    use self::jni::sys::{jstring};
+
+    #[no_mangle]
+    pub unsafe extern fn Java_com_mozilla_greetings_RustGreetings_greeting(env: JNIEnv, _: JClass, java_pattern: JString) -> jstring {
+        let world = rust_greeting(env.get_string(java_pattern).expect("invalid pattern string").as_ptr());
+        let output = env.new_string(CStr::from_ptr(world).to_str().unwrap()).expect("Couldn't create java string!");
+        rust_greeting_free(world);
+
+        output.into_inner()
+    }
+}
+```
+The first line here `#[cfg(target_os="android")]` is telling the compiler to target Android when compiling this module. `#[cfg]` is a special attribute that allows you to compile code based on a flag passed to the compiler. 
+
+The second line, `#[allow(non_snake_case)]`, tells the compiler not to warn if we are not using snake_case for a variable or function name. The Rust compiler is very strict, this is one of the things that makes Rust great - it's nearly impossible to code something that is not correct, and it enforces the use of snake_case throughout. However, we defined our class name and native method in our Android project using Java coding conventions which is camelCase and UpperCamelCase and we don't want to change this or our Java code will look wrong. Given the way that JNI constructs native function names, we need to tell the Rust compiler to go easy on us in this instance. This flag will apply to all functions and variables created inside this module that we are creating, called `android`. 
+
+After declaring that we need the `jni` create and importing some things from it, we can declare our function. This function needs to be marked `unsafe` because we will be dealing with pointers from a language that allows null pointers, but our code doesn't check for `NULL`. This situation would never happen in strictly Rust code as the Rust compiler enforces memory safety. By marking the function as not memory safe, we are alerting other Rust functions that it may not be able to deal with a null pointer. `extern` defines the function as one that will be exposed to other languages. 
```

Comment:
First sentence is missing a word or two.

By "strictly" I think you mean Rust-only, but you might mean "strict" or "safe"?